benches/engines/mod.rs

```diff
@@ -0,0 +1,232 @@
+    engine: &E,
+    write_count: usize,
+    value_length: usize,
+) {
+    let test_kvs = generate_random_kvs(write_count, DEFAULT_KEY_LENGTH, value_length);
+    let ctx = Context::new();
+    bencher.iter(|| {
+        for (key, value) in &test_kvs {
+            let _ = engine.put(&ctx, Key::from_raw(&key), value.clone());
+        }
+    });
+}
+
+#[derive(Debug)]
+struct PutConfig {
+    engine_type: EngineType,
+    put_count: usize,
+    value_length: usize,
+}
+
+fn bench_engine_put(bencher: &mut Bencher, config: &PutConfig) {
+    match config.engine_type {
+        EngineType::BTreeEngine => {
+            let engine = BTreeEngine::default();
+            engine_put_bench(bencher, &engine, config.put_count, config.value_length)
+        }
+        EngineType::RocksDB => {
+            let engine = TestEngineBuilder::new().build().unwrap();
+            engine_put_bench(bencher, &engine, config.put_count, config.value_length)
+        }
+    }
+}
+
+#[derive(Debug)]
+struct SnapshotConfig {
+    engine_type: EngineType,
+    engine_keys_count: usize,
+    value_length: usize,
+}
+
+fn bench_engine_snapshot(bencher: &mut Bencher, config: &SnapshotConfig) {
+    match config.engine_type {
+        EngineType::BTreeEngine => {
+            let engine = BTreeEngine::default();
+            engine_snapshot_bench(
+                bencher,
+                &engine,
+                config.engine_keys_count,
+                config.value_length,
+            )
+        }
+        EngineType::RocksDB => {
+            let engine = TestEngineBuilder::new().build().unwrap();
+            engine_snapshot_bench(
+                bencher,
+                &engine,
+                config.engine_keys_count,
+                config.value_length,
+            )
+        }
+    }
+}
+
+#[derive(Debug)]
+struct GetConfig {
+    engine_type: EngineType,
+    iterations: usize,
+    value_length: usize,
+    engine_keys_count: usize,
+}
+
+fn bench_engine_get(bencher: &mut Bencher, config: &GetConfig) {
+    match config.engine_type {
+        EngineType::BTreeEngine => {
+            let engine = BTreeEngine::default();
+            engine_get_bench(
+                bencher,
+                &engine,
+                config.iterations,
+                config.engine_keys_count,
+                config.value_length,
+            )
+        }
+        EngineType::RocksDB => {
+            let engine = TestEngineBuilder::new().build().unwrap();
+            engine_get_bench(
+                bencher,
+                &engine,
+                config.iterations,
+                config.engine_keys_count,
+                config.value_length,
+            )
+        }
+    }
+}
+
+fn bench_engines(c: &mut Criterion) {
+    let engine_types = vec![EngineType::BTreeEngine, EngineType::RocksDB];
+    let value_lengths = vec![128, 1024];
+    let engine_keys_counts = vec![0, 1000, 10_000];
+    let engine_put_keys_counts = vec![0, 1000];
```

Comment:
what does "put 0 keys" mean?

Reply:
It was used to quick bench for my test where all counts are zeros, bur I forgot to remove it at last.😂